src/alpaka/AlpakaCore/CachingDeviceAllocator.h

```diff
@@ -0,0 +1,444 @@
+               (long long)this->max_cached_bytes,
+               (long long)max_cached_bytes);
+      }
+
+      this->max_cached_bytes = max_cached_bytes;
+
+      mutex_locker.unlock();
+    }
+
+    /**
+     * \brief Provides a suitable allocation of device memory for the given size on the specified device.
+     *
+     * Once freed, the allocation becomes available immediately for reuse.
+     */
+    auto DeviceAllocate(size_t bytes,  ///< [in] Minimum no. of bytes for the allocation
+                        const ::ALPAKA_ACCELERATOR_NAMESPACE::Device&
+                            device)  ///< [in] The device to be associated with this allocation
+    {
+      std::unique_lock<std::mutex> mutex_locker(mutex, std::defer_lock);
+      int device_idx = getIdxOfDev(device);
+
+      // Create a block descriptor for the requested allocation
+      bool found = false;
+      auto [bin, bin_bytes] = NearestPowerOf(bin_growth, bytes);
+      BlockDescriptor search_key{bin, bin_bytes, bytes, device};
+
+      if (search_key.bin > max_bin) {
+        // Bin is greater than our maximum bin: allocate the request
+        // exactly and give out-of-bounds bin.  It will not be cached
+        // for reuse when returned.
+        search_key.bin = INVALID_BIN;
+        search_key.bytes = bytes;
+      } else {
+        // Search for a suitable cached allocation: lock
+        mutex_locker.lock();
+
+        if (search_key.bin < min_bin) {
+          // Bin is less than minimum bin: round up
+          search_key.bin = min_bin;
+          search_key.bytes = min_bin_bytes;
+        }
+
+        // Find a cached block on the same device in the same bin
+        auto block_itr = cached_blocks.find(search_key);
+        if (block_itr != cached_blocks.end()) {
+          // Reuse existing cache block.  Insert into live blocks.
+          found = true;
+          search_key = *block_itr;
+          live_blocks.insert(search_key);
+
+          // Remove from free blocks
+          cached_bytes[device_idx].free -= search_key.bytes;
+          cached_bytes[device_idx].live += search_key.bytes;
+          cached_bytes[device_idx].liveRequested += search_key.bytesRequested;  // CMS
+
+          if (debug) {
+            // CMS: improved debug message
+            // CMS: use raw printf
+            printf("\tDevice %d reused cached block at %p (%lld bytes).\n",
+                   device_idx,
+                   alpaka::getPtrNative(search_key.buf),
+                   (long long)search_key.bytes);
+          }
+
+          cached_blocks.erase(block_itr);
+        }
+        // Done searching: unlock
+        mutex_locker.unlock();
+      }
+
+      // Allocate the block if necessary
+      if (!found) {
+        search_key.buf = alpaka::allocBuf<std::byte, alpaka_common::Idx>(
+            device, static_cast<alpaka_common::Extent>(search_key.bytes));
+#if CUDA_VERSION >= 11020
+        alpaka::prepareForAsyncCopy(search_key.buf);
```

Comment:
What does `prepareForAsyncCopy()` do? I'd guess it to call `cudaHostRegister()` or similar under the hood for host memory.

Looking at the code I see this is a no-op for CUDA/HIP buffers
https://github.com/alpaka-group/alpaka/blob/ee525dbb1c3e71490a17df80e7d13b3067619b95/include/alpaka/mem/buf/BufUniformCudaHipRt.hpp#L408-L418
(so in principle this call would not be needed for the device allocator)

For host buffers it indeed ends up calling `cudaHostRegister()`
https://github.com/alpaka-group/alpaka/blob/ee525dbb1c3e71490a17df80e7d13b3067619b95/include/alpaka/mem/buf/BufCpu.hpp#L363-L378
https://github.com/alpaka-group/alpaka/blob/ee525dbb1c3e71490a17df80e7d13b3067619b95/include/alpaka/mem/buf/BufCpu.hpp#L269-L303

But given the
```cpp
#if(defined(ALPAKA_ACC_GPU_CUDA_ENABLED) && BOOST_LANG_CUDA) || (defined(ALPAKA_ACC_GPU_HIP_ENABLED) && BOOST_LANG_HIP)
```
the `cudaHostRegister()` is called only if the source file including this Alpaka header is compiled with `ALPAKA_ACC_GPU_CUDA_ENABLED` and compiling with `nvcc` (which we do when that macro is enabled). I'm afraid mixing that with code using `BufCpu.hpp` compiled without `ALPAKA_ACC_GPU_CUDA_ENABLED` would technically violate ODR.

I also see the definition of `alpaka::detail::BufCpuImpl` depends on these macros, in particular whether the `m_bPinned` member exists or not
https://github.com/alpaka-group/alpaka/blob/ee525dbb1c3e71490a17df80e7d13b3067619b95/include/alpaka/mem/buf/BufCpu.hpp#L103-L105